### PR TITLE
Add support for partial feature spec extraction

### PR DESCRIPTION
**Pull request by regadas**

### CODE REVIEW

File: core/src/main/scala/com/spotify/featran/FeatureExtractor.scala [hunk 1]
```diff
@@ -40,17 +40,28 @@ class FeatureExtractor[M[_]: CollectionType, T] private[featran] (
 
   @transient private[featran] lazy val as: M[(T, ARRAY)] = {
     val g = fs // defeat closure
-    input.map(o => (o, g.unsafeGet(o)))
+    input.cross(g).map { case (in, spec) => (in, spec.unsafeGet(in)) }
```
Comment (nevillelyh) [inline core/src/main/scala/com/spotify/featran/FeatureExtractor.scala]: Are these closure => `cross` changes to defeat closure serialization?

File: core/src/main/scala/com/spotify/featran/FeatureExtractor.scala [hunk 2]
```diff
@@ -90,14 +102,14 @@ class FeatureExtractor[M[_]: CollectionType, T] private[featran] (
    *           `DenseVector[Double]`
    */
   def featureResults[F: FeatureBuilder: ClassTag]: M[FeatureResult[F, T]] = {
-    val fb = CrossingFeatureBuilder(implicitly[FeatureBuilder[F]].newBuilder, fs.crossings)
-    as.cross(aggregate).map {
-      case ((o, a), c) =>
-        fs.featureValues(a, c, fb)
-        FeatureResult(fb.result, fb.rejections, o)
+    val fb = implicitly[FeatureBuilder[F]].newBuilder
+    as.cross(aggregate).cross(fs).map {
```
Comment (nevillelyh) [inline core/src/main/scala/com/spotify/featran/FeatureExtractor.scala]: `as.cross((aggregate, fs))` to remove an extra side input operation?

File: core/src/main/scala/com/spotify/featran/FeatureSpec.scala [hunk 3]
```diff
@@ -109,8 +109,62 @@ class FeatureSpec[T] private[featran] (private[featran] val features: Array[Feat
...
+  /**
+   * Extract features from an input collection based on the provided predicate
+   *
+   * @param input input collection
+   * @param predicate Function determining whether or not to include the feature
+   * @tparam M input collection type, e.g. `Array`, `List`
+   */
+  def extract[M[_]: CollectionType](
+    input: M[T],
+    predicate: Feature[T, _, _, _] => Boolean): FeatureExtractor[M, T] = {
```
Comment (nevillelyh) [inline core/src/main/scala/com/spotify/featran/FeatureSpec.scala]: Curry `predicate` so one can do `.extract(in) { f => ... }`?

File: core/src/main/scala/com/spotify/featran/MultiFeatureSpec.scala [hunk 4]
```diff
@@ -47,13 +49,39 @@ class MultiFeatureSpec[T](private[featran] val mapping: Map[String, Int],
...
+  /**
+   * Extract features from an input collection based on the provided predicate
+   *
+   * @param input input collection
+   * @param predicate Function determining whether or not to include the feature
+   * @tparam M input collection type, e.g. `Array`, `List`
+   */
+  def extract[M[_]: CollectionType](
+    input: M[T],
+    predicate: Feature[T, _, _, _] => Boolean): MultiFeatureExtractor[M, T] = {
```
Comment (nevillelyh) [inline core/src/main/scala/com/spotify/featran/MultiFeatureSpec.scala]: Curry `predicate`

**Comment by nevillelyh** [comment]
You need to rebase on `master` now that #123 is merged.

File: core/src/main/scala/com/spotify/featran/FeatureSpec.scala (hunk 3 above)
Comment (regadas) [reply core/src/main/scala/com/spotify/featran/FeatureSpec.scala]: @nevillelyh thought about that as well and then i was torn.. coz if we curry it we need to rename that method.

`extractWithPredicate`?

File: core/src/main/scala/com/spotify/featran/FeatureExtractor.scala (hunk 2 above)
Comment (regadas) [reply core/src/main/scala/com/spotify/featran/FeatureExtractor.scala]: @nevillelyh so this is where it gets tricky ... `fs: FeatureSet[T]` is now `fs: M[FeatureSet[T]]`

File: core/src/main/scala/com/spotify/featran/FeatureSpec.scala (hunk 3 above)
Comment (nevillelyh) [reply core/src/main/scala/com/spotify/featran/FeatureSpec.scala]: Yeah that could work, or `extractSubset`?

Comment (regadas) [reply core/src/main/scala/com/spotify/featran/FeatureSpec.scala]: `extractSubset`

File: core/src/main/scala/com/spotify/featran/FeatureSpec.scala [hunk 5]
```diff
@@ -122,8 +172,34 @@ class FeatureSpec[T] private[featran] (private[featran] val features: Array[Feat
...
+  /**
+   * Extract features from individual records using partial settings. Since the
+   * settings are parsed only once, this is more efficient and is recommended when the input is
+   * from an unbounded source, e.g. a stream of events or a backend service.
+   *
+   * This bypasses the `reduce` step in [[extract]] and uses feature summary from settings exported
+   * in a previous session.
+   * @param settings JSON settings from a previous session
+   */
+  def extractWithPartialSettings[F: FeatureBuilder: ClassTag](
```
Comment (regadas) [inline core/src/main/scala/com/spotify/featran/FeatureSpec.scala]: @nevillelyh should we rename this to `extractWithSettingsSubset`?

Comment (regadas) [reply core/src/main/scala/com/spotify/featran/FeatureSpec.scala]: renamed it to `extractWithSubSetSettings`

File: core/src/main/scala/com/spotify/featran/FeatureSpec.scala [hunk 6]
```diff
@@ -108,8 +110,45 @@ class FeatureSpec[T] private[featran] (private[featran] val features: Array[Feat
...
+    new FeatureExtractor[M, T](fs, input, None)
+  }
+
+  /**
+   * Creates a new FeatureSpec with only the features that respect the given predicate.
+   *
+   * @param predicate Function determining whether or not to include the feature
+   */
+  def filter(predicate: Feature[T, _, _, _] => Boolean): FeatureSpec[T] =
+    new FeatureSpec[T](features.filter(predicate), crossings)
```
Comment (richwhitjr) [inline core/src/main/scala/com/spotify/featran/FeatureSpec.scala]: Should we be filtering crossings here and other places?

Comment (regadas) [reply core/src/main/scala/com/spotify/featran/FeatureSpec.scala]: @richwhitjr indeed! I missed even after were talking about it lol!

Comment (regadas) [reply core/src/main/scala/com/spotify/featran/FeatureSpec.scala]: @richwhitjr I think this is what we need right?

**Pull request merged**